### PR TITLE
issue1539 durack1 update iPython 3.0.0 to 4.1.2

### LINKED ISSUE #1539: Update iPython to 4.1.2
We are a full release behind iPython latest (3.0.0 vs 4.0.0).

@doutriaux1 want to assign me as a 2.4 milestone?

### PR DESCRIPTION
**Pull request by durack1**
Fix #1539 - Update iPython 3.0.0 to 4.0.0 with dependencies

@doutriaux1 please review and merge

### CODE REVIEW

**Comment by durack1** [comment]
@doutriaux1 it would be good to add a test for this..

**Comment by durack1** [comment]
@doutriaux1 can you also check the dependencies - not sure whether `ipython` should be a dependency of the 2 new packages?

**Comment by dnadeau4** [comment]
Add this in one of the testing directory into CMakeLists.txt
[BUILD]/testing
`add_test(run_ipython ${CMAKE_INSTALL_PREFIX}/bin/ipython -V  )`

**Comment by aashish24** [comment]
-1, @durack1 let's take care of that post 2.4

**Comment by durack1** [comment]
@aashish24 @doutriaux1 2.4+ is fine with me.. There are a bunch of additional dependencies that are still required, the `ipython2 -> 3 -> 4` has changed things considerably..

**Comment by durack1** [comment]
@doutriaux1 a low priority, but this is now good to go..
```
*******************************
UVCDAT  - issue1539_durack1_UpdateIpython3.0.0To4.0.0 - Install Success
Metrics - master - Install Success
*******************************
```

I'll re-test using `num_cpus=1` (I used 4) just to make sure that there are no dependencies that are likely to trip things over - otherwise it's good to go..

**Comment by durack1** [comment]
FYI @doutriaux1 `num_cpus=1`=
```
*******************************
UVCDAT  - master - Install Success
Metrics - master - Install Success
*******************************
```

**Comment by durack1** [comment]
@doutriaux1 with all my testing on `ocean` it looks like this is ready for testing across the suite.. When you can wedge it in, please do test, review and merge

**Comment by doutriaux1** [comment]
building on my macs

**Comment by durack1** [comment]
@doutriaux1 perfect, let me know how it goes.. Also the `ipython` test needs tweaking - [it failed](https://open.cdash.org/viewTest.php?onlyfailed&buildid=4262754) along with some other unrelated `VCS` tests

File: CMake/cdat_modules/pyexpect_pkg.cmake [hunk 1]
```diff
@@ -0,0 +1,10 @@
+set(PYEXPECT_MAJOR 4)
+set(PYEXPECT_MINOR 0)
+set(PYEXPECT_PATCH 1)
+set(PYEXPECT_VERSION ${PYEXPECT_MAJOR}.${PYEXPECT_MINOR}.${PYEXPECT_PATCH})
+set(PYEXPECT_URL ${LLNL_URL} )
+set(PYEXPECT_GZ pyexpect-${PYEXPECT_VERSION}.tar.gz)
+set(PYEXPECT_MD5 2bd260f7f2159f9bcab373721736d526)
+set(PYEXPECT_SOURCE ${PYEXPECT_URL}/${PYEXPECT_GZ})
+
+add_cdat_package(IPYTHON "" "" ON)
```
Comment (durack1) [inline CMake/cdat_modules/pyexpect_pkg.cmake]: @doutriaux1 your traitlets fix means this is also incorrect too - so changing to PYEXPECT (from IPYTHON)

**Comment by doutriaux1** [comment]
@durack1 what about
```
File "/lgm/uvcdat/2016-03-04/Library/Frameworks/Python.framework/Versions/2.7/bin/ipython", line 4, in <module>
    from IPython import start_ipython
  File "/lgm/uvcdat/2016-03-04/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/__init__.py", line 48, in <module>
    from .core.application import Application
  File "/lgm/uvcdat/2016-03-04/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/IPython/core/application.py", line 22, in <module>
    from traitlets.config.application import Application, catch_config_error
  File "/lgm/uvcdat/2016-03-04/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/traitlets/__init__.py", line 1, in <module>
    from .traitlets import *
  File "/lgm/uvcdat/2016-03-04/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/traitlets/traitlets.py", line 56, in <module>
    from ipython_genutils import py3compat
ImportError: No module named ipython_genutils
```

**Comment by durack1** [comment]
@doutriaux1 peel - https://pypi.python.org/pypi/ipython_genutils

**Comment by doutriaux1** [comment]
I know it's insane! But that's why I don't care so much for pip so much stuff brought in w/o the user knowing about it!

**Comment by durack1** [comment]
@doutriaux1 ok let's see how `travis` deals with this.. If it's a green light then can you test `ALL` and `LEAN` - I'm not sure about the `tornado` dependency difference between `ALL` and `LEAN` in `ipython_deps`

File: testing/CMakeLists.txt [hunk 2]
```diff
@@ -46,6 +46,9 @@ if (CDAT_BUILD_MATPLOTLIB)
   add_subdirectory(matplotlib)
 endif()
 
+# Test ipython
+add_test(run_ipython ${CMAKE_INSTALL_PREFIX}/bin/ipython -V)
```
Comment (durack1) [inline testing/CMakeLists.txt]: @doutriaux1 you might want to tweak this "test".. I'm not sure how the return codes are dealt with, and whether this is actually an effective "test"

**Comment by durack1** [comment]
@doutriaux1 there's a [build order issue here, or a dependency inconsistency](https://travis-ci.org/UV-CDAT/uvcdat/jobs/113821085):
```
...
[INFO] Package --- PYEXPECT/PYEXPECT/4.0.1 --- will be built
[INFO] Package --- TRAITLETS/TRAITLETS/4.0.0 --- will be built
[INFO] Package --- IPYTHON_GENUTILS/IPYTHON_GENUTILS/0.1.0 --- will be built
[INFO] Package --- IPYTHON/IPYTHON/4.1.2 --- will be built
[INFO] ADDITIONAL CFLAGS 
[INFO] [PIP] Installing IPYTHON from tarball
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET PYEXPECT.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:647 (include)
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET TRAITLETS.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:647 (include)
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET IPYTHON_GENUTILS.  Perhaps it has not
  yet been created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:647 (include)
```

**Comment by doutriaux1** [comment]
will take a look, but probably on monday, sorry...

**Comment by durack1** [comment]
@doutriaux1 no problem, it's a low priority anyway.. It's close I think, but requires some magic syntax tweaking..

**Comment by durack1** [comment]
@doutriaux1 @aashish24 not sure if you folks know what is going on here in the [travis builds](https://travis-ci.org/UV-CDAT/uvcdat/jobs/114898685):
```
...
[INFO] Package --- mccabe/MCCABE/0.3.1 --- will be built
[INFO] Package --- flake8/FLAKE8/2.4.1 --- will be built
[INFO] Package --- PYEXPECT/PYEXPECT/4.0.1 --- will be built
[INFO] Package --- TRAITLETS/TRAITLETS/4.0.0 --- will be built
[INFO] Package --- IPYTHON_GENUTILS/IPYTHON_GENUTILS/0.1.0 --- will be built
[INFO] Package --- IPYTHON/IPYTHON/4.1.2 --- will be built
[INFO] ADDITIONAL CFLAGS 
[INFO] [PIP] Installing IPYTHON from tarball
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET PYEXPECT.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET TRAITLETS.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET IPYTHON_GENUTILS.  Perhaps it has not
  yet been created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
```
These libraries are included in the `ipython_deps.cmake` file:
```
  if (CDAT_BUILD_ALL)
+  set(IPYTHON_deps ${pip_pkg} ${numpy_pkg} ${pyexpect_pkg} ${traitlets_pkg} ${ipython_genutils_pkg} ${tornado_pkg} ${numexpr_pkg})
 else ()
+  set(IPYTHON_deps ${pip_pkg} ${numpy_pkg} ${pyexpect_pkg} ${traitlets_pkg} ${ipython_genutils_pkg})
 endif()
```

**Comment by durack1** [comment]
@aashish24 @doutriaux1 can you provide some guidance here - I'm not sure what is going wrong:
```
[INFO] Package --- pep8/PEP8/1.5.7 --- will be built
[INFO] Package --- mccabe/MCCABE/0.3.1 --- will be built
[INFO] Package --- flake8/FLAKE8/2.4.1 --- will be built
[INFO] Package --- PYEXPECT/PYEXPECT/4.0.1 --- will be built
.[INFO] Package --- TRAITLETS/TRAITLETS/4.0.0 --- will be built
[INFO] Package --- IPYTHON_GENUTILS/IPYTHON_GENUTILS/0.1.0 --- will be built
[INFO] Package --- IPYTHON/IPYTHON/4.1.2 --- will be built
[INFO] ADDITIONAL CFLAGS 
[INFO] [PIP] Installing IPYTHON from tarball
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET PYEXPECT.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
.CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET TRAITLETS.  Perhaps it has not yet been
  created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
CMake Error at CMake/cmake_modules/ExternalProject.cmake:1397 (get_property):
  get_property could not find TARGET IPYTHON_GENUTILS.  Perhaps it has not
  yet been created.
Call Stack (most recent call first):
  CMake/cmake_modules/ExternalProject.cmake:1600 (_ep_add_configure_command)
  CMake/cmake_modules/pipinstaller.cmake:26 (ExternalProject_Add)
  CMake/cdat_modules/ipython_external.cmake:6 (include)
  CMakeLists.txt:648 (include)
[INFO] Package --- sampledata/SAMPLEDATA/ --- will be built
[INFO] Package --- singledispatch/SINGLEDISPATCH/3.4.0.3 --- will be built
```

**Comment by doutriaux1** [comment]
i will take a look, probably some name typo or cicrular dependency

**Comment by durack1** [comment]
@doutriaux1 thanks, it's not obvious to me at all..

**Comment by aashish24** [comment]
@durack1 could you please resolve the conflict?

**Comment by durack1** [comment]
@aashish24 would you mind taking a look at this, I'm not sure what is wrong with it

**Comment by durack1** [comment]
@aashish24 ping!

**Comment by aashish24** [comment]
> @aashish24 would you mind taking a look at this, I'm not sure what is wrong with it

Sure, I will @durack1

**Comment by durack1** [comment]
@aashish24 great! I'd be intrigued to know what it is/was.. The `CMake` Magic baffles me

**Comment by aashish24** [comment]
@durack1 its actually quite simple. PYEXPECT is not an target meaning that CMake does not know anything about it but for your package it is required so it seems like either a order issue or missing dependencies issue. 

Like make target, cmake has targets too.

**Comment by aashish24** [comment]
> @durack1 its actually quite simple. PYEXPECT is not an target meaning that CMake does not know anything about it but for your package it is required so it seems like either a order issue or missing dependencies issue.

Obviously the hard part is what went wrong -:)

**Comment by durack1** [comment]
@aashish24 `Pyexpect_pkg` is included in `CMakeLists.txt` so I thought the PR was complete

**Comment by aashish24** [comment]
@durack1 I fixed all the issues I believe. Basically, the external project add is case sensitive. It has to match with the package name that is listed under _pkg.cmake file.

**Comment by durack1** [comment]
@aashish24, case sensitive? Is this OS dependent? Great, looking forward to seeing this one merged

**Comment by aashish24** [comment]
> @aashish24, case sensitive? Is this OS dependent?

no, not OS dependent. The Cmake external project is case sensitive so if you say

`ExternalProject_Add(Foo`

is different than

`ExternalProject_Add(foo`

In the first case targe foo won't be found which was referred in the dependencies of other package.

**Comment by aashish24** [comment]
I am going to merge this one since travis is passing (garant is stuck) but this branch should be very safe from testing point of view.

**Comment by durack1** [comment]
@aashish24 if you read the inline comments, I do think the cursory test needs checking.. I'm unfamiliar with how `ctest` works with return codes etc

**Comment by aashish24** [comment]
> I do think the cursory test needs checking..

Where is that? Do you mean the the ipython test?

File: testing/CMakeLists.txt [hunk 3]
```diff
@@ -55,6 +55,9 @@ if (CDAT_BUILD_MATPLOTLIB)
...
```
Comment (durack1) [inline testing/CMakeLists.txt]: @aashish24 this is what I was referring to - not certain it works as I'd guessed

**Comment by durack1** [comment]
@aashish24 take a look at the file `testing/CMakeLists.txt` and [my comment](https://github.com/UV-CDAT/uvcdat/pull/1675/files#r57325558)

File: testing/CMakeLists.txt (hunk 3 above)
Comment (aashish24) [inline testing/CMakeLists.txt]: @durack1 it should work, if the command is not found, I believe shell will set the error flag and ctest would know that test has failed. I can do local testing to make sure this works as expected but my initial reaction was that it should and that's why I approved the branch.

Comment (durack1) [inline testing/CMakeLists.txt]: Ok great, I had tested this outside of `ctest` however wasn't sure..

**Comment by doutriaux1** [comment]
@aashish24 it DOES NOT WORK....

I get:

```
Performing build step for 'PYEXPECT'
[  6%] No update step for 'pkgconfig'
[  6%] [  6%] [  6%] [  6%] CMake Error at /home/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build.cmake:56 (message):
  CTEST_FULL_OUTPUT

   Command failed: No such file or directory
   '/lgm/uvcdat/2016-03-24/bin/python' 'setup.py' 'build'
   From /home/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build-out.log

  -------------------------------------------------

  -------------------------------------------------

   From /home/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build-err.log

  -------------------------------------------------

  -------------------------------------------------



CMakeFiles/PYEXPECT.dir/build.make:107: recipe for target 'PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build' failed
make[2]: *** [PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build] Error 1
CMakeFiles/Makefile2:1894: recipe for target 'CMakeFiles/PYEXPECT.dir/all' failed
make[1]: *** [CMakeFiles/PYEXPECT.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
Downloading acme_hires_land/c_t_t341f02.F1850p.clm2.h0.1852-07.nc
```

**Comment by aashish24** [comment]
> @aashish24 it DOES NOT WORK....
> 
> I get:

@doutriaux1 please do not revert based on your machine failures. The dashboards passed (all three of them) before we merged this branch. Also, it sounds like your package didnt' download (network issue may be?). 

Also, are you running on supported platforms or Ubuntu 15.* which is not officially supported.

**Comment by doutriaux1** [comment]
Yes ubuntu is officially supported.

**Comment by doutriaux1** [comment]
Anyhow, I reverted because if it breaks on my machine it's definitely going to break on many machines. You should have waited to hear from me, even @durack1 was having reservation about merging just yet.
Same for OCGIS you ask me to check your changes and then merge before I get back to you. This is not standard practice @aashish24

**Comment by aashish24** [comment]
> even @durack1 was having reservation about merging just yet

he was because of the build issue. 

> This is not standard practice @aashish24

@doutriaux1 I am happy to talk about the standard practice. Both of these branches were not mine, I fixed the issue, the dashboards passed (I built them locally) and that's why I merged. If they were mine, I would not have merged it. On both of these branches I was asked to look into the issue. 

> Anyhow, I reverted because if it breaks on my machine it's definitely going to break on many machines.

What is your machine running? Like I said it passes on travis, mac, and ubuntu 14.04.

**Comment by doutriaux1** [comment]
@aashish24 from my mac:
```
[  8%] Downloading acme_hires_land/c_t_t341f02.F1850p.clm2.h0.1853-01.nc
CMake Error at /Users/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build.cmake:56 (message):
  CTEST_FULL_OUTPUT

   Command failed: No such file or directory
   '/lgm/uvcdat/ipython/bin/python' 'setup.py' 'build'
   From /Users/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build-out.log

  -------------------------------------------------

  -------------------------------------------------

   From /Users/doutriaux1/build/PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build-err.log

  -------------------------------------------------

  -------------------------------------------------



make[2]: *** [PYEXPECT-prefix/src/PYEXPECT-stamp/PYEXPECT-build] Error 1
```

**Comment by doutriaux1** [comment]
there's obviously something broken...

**Comment by doutriaux1** [comment]
beside you wait for months to look at this, we are on the verge of tagging a release and you suddenly wake up and start merging lots of branches that are absolutely not relevant... Very frustrating

**Comment by aashish24** [comment]
@doutriaux1 that's the point. we have these PR sittign for long time and I was trying to clean up. R worked on travis CI and on my systems and that' why I merged it.  It looks like your system for some reason its not building correctly.Also the issue you are having right now is not something you got before. 

on ubuntu, my question was yes we support ubuntu but does we support Ubuntu 15.10? We don't support every version of Ubuntu do we?

**Comment by aashish24** [comment]
@doutriaux1 I am looking into your build failure on new VM.

**Pull request merged**